runner/providers/external/external.go

```diff
@@ -2,334 +2,22 @@ package external
-
-	metrics.InstanceOperationCount.WithLabelValues(
-		"RemoveAllInstances", // label: operation
-		e.cfg.Name,           // label: provider
-	).Inc()
-
-	_, err := garmExec.Exec(ctx, e.execPath, nil, asEnv)
-	if err != nil {
-		metrics.InstanceOperationFailedCount.WithLabelValues(
-			"RemoveAllInstances", // label: operation
-			e.cfg.Name,           // label: provider
-		).Inc()
-		return garmErrors.NewProviderError("provider binary %s returned error: %s", e.execPath, err)
-	}
-	return nil
-}
-
-// Stop shuts down the instance.
-func (e *external) Stop(ctx context.Context, instance string) error {
-	asEnv := []string{
-		fmt.Sprintf("GARM_COMMAND=%s", execution.StopInstanceCommand),
-		fmt.Sprintf("GARM_CONTROLLER_ID=%s", e.controllerID),
-		fmt.Sprintf("GARM_INSTANCE_ID=%s", instance),
-		fmt.Sprintf("GARM_PROVIDER_CONFIG_FILE=%s", e.cfg.External.ConfigFile),
-	}
-	asEnv = append(asEnv, e.environmentVariables...)
-
-	metrics.InstanceOperationCount.WithLabelValues(
-		"Stop",     // label: operation
-		e.cfg.Name, // label: provider
-	).Inc()
-	_, err := garmExec.Exec(ctx, e.execPath, nil, asEnv)
-	if err != nil {
-		metrics.InstanceOperationFailedCount.WithLabelValues(
-			"Stop",     // label: operation
-			e.cfg.Name, // label: provider
-		).Inc()
-		return garmErrors.NewProviderError("provider binary %s returned error: %s", e.execPath, err)
-	}
-	return nil
-}
-
-// Start boots up an instance.
-func (e *external) Start(ctx context.Context, instance string) error {
-	asEnv := []string{
-		fmt.Sprintf("GARM_COMMAND=%s", execution.StartInstanceCommand),
-		fmt.Sprintf("GARM_CONTROLLER_ID=%s", e.controllerID),
-		fmt.Sprintf("GARM_INSTANCE_ID=%s", instance),
-		fmt.Sprintf("GARM_PROVIDER_CONFIG_FILE=%s", e.cfg.External.ConfigFile),
-	}
-	asEnv = append(asEnv, e.environmentVariables...)
-
-	metrics.InstanceOperationCount.WithLabelValues(
-		"Start",    // label: operation
-		e.cfg.Name, // label: provider
-	).Inc()
-
-	_, err := garmExec.Exec(ctx, e.execPath, nil, asEnv)
-	if err != nil {
-		metrics.InstanceOperationFailedCount.WithLabelValues(
-			"Start",    // label: operation
-			e.cfg.Name, // label: provider
-		).Inc()
-		return garmErrors.NewProviderError("provider binary %s returned error: %s", e.execPath, err)
-	}
-	return nil
-}
-
-func (e *external) AsParams() params.Provider {
-	return params.Provider{
-		Name:         e.cfg.Name,
-		Description:  e.cfg.Description,
-		ProviderType: e.cfg.ProviderType,
-	}
-}
-
-// DisableJITConfig tells us if the provider explicitly disables JIT configuration and
-// forces runner registration tokens to be used. This may happen if a provider has not yet
-// been updated to support JIT configuration.
-func (e *external) DisableJITConfig() bool {
-	if e.cfg == nil {
-		return false
+	switch cfg.External.InterfaceVersion {
+	case "v0.1.0":
+		return v010.NewProvider(ctx, cfg, controllerID)
+	case "v0.1.1":
+		return v011.NewProvider(ctx, cfg, controllerID)
+	default:
```

Comment:
on default we should probably err out. We should default empty string to `0.1.0` though.